File: scala/HeapTopK.scala
import collection.mutable
//import scala.math.Ordering.Implicits._

object HeapTopK {
   // get K biggest elements of the array (or K-th biggest)
   // O(n log n) is general case compelxity
   def heapMaxTopK(arr: Array[Int], k: Int): Array[Int] = {
        val heap: mutable.PriorityQueue[Int] = mutable.PriorityQueue.from(arr)
         
        val ab = mutable.ArrayBuffer[Int]()
        for (i<-0 until k) {
           ab += heap.dequeue()
        }
        ab.toArray
   }

   object MinOrder extends Ordering[Int] {
        def compare(x:Int, y:Int) = y compare x
    }

    // get K smallest elements of the array (or K-th smallest)
    // O(n log n) is general case compelxity
    def heapMinTopK(arr: Array[Int], k: Int): Array[Int] = {
        val heap: mutable.PriorityQueue[Int] = mutable.PriorityQueue.from(arr)(MinOrder)

        val ab = mutable.ArrayBuffer[Int]()
        for (i<-0 until k) {
            ab += heap.dequeue()
        }
        ab.toArray
   }

    //scalac HeapTopK.scala && scala HeapTopK
    def main(args: Array[String])={
        val array = Array(1,7,2,6,3,5,4)
        println(s"3 highest numbers of array are ${heapMaxTopK(array, 3).mkString(",")} [7,6,5]")
        println(s"4 lowest numbers of array are ${heapMinTopK(array, 4).mkString(",")} [1,2,3,4]")
    }
}

File: java/HeapTopK.java
import java.util.*;
import java.util.stream.*;

class HeapTopK {
    // get K biggest or smallest elements of the array (or K-th biggest)
    // O(n log n) is general case compelxity
    private static int[] heapTopK(int[] arr, int k, boolean isMax) {
        List<Integer> boxed = IntStream.of(arr).boxed().toList();
        var heap = new PriorityQueue<Integer>((x, y) -> isMax? Integer.compare(y, x): Integer.compare(x,y));
        heap.addAll(boxed);
        ArrayList<Integer> ab = new ArrayList<Integer>();
        for (int i=0; i<k; i++) {
            ab.add(heap.poll());
        }
        return ab.stream().mapToInt(i -> i).toArray();
    }

    // javac HeapTopK.java && java HeapTopK 
    public static void main(String[] args) {
        int[] array = new int[]{1,7,2,6,3,5,4};
        System.out.printf("3 highest numbers of array are %s [7,6,5]\n", Arrays.toString(heapTopK(array, 3, true)));
        System.out.printf("4 lowest numbers of array are %s [1,2,3,4]\n",  Arrays.toString(heapTopK(array, 4, false)));
    }
}